File: examples/csharp/Player/Agg/Handlers/RegisterHandler.cs
using Angzarr.Client;
using Angzarr.Examples;
using Google.Protobuf.WellKnownTypes;

namespace Player.Agg.Handlers;

/// <summary>
/// Handler for RegisterPlayer command.
///
/// Follows the guard/validate/compute pattern:
/// - Guard: Check state preconditions (aggregate exists, correct phase)
/// - Validate: Validate command inputs
/// - Compute: Build the resulting event
///
/// Why this pattern? Each step is a pure function (state in, result out),
/// enabling direct unit testing without mocking infrastructure. You can test
/// each step independently by passing state objects and asserting on results.
/// </summary>
public static class RegisterHandler
{
    public static PlayerRegistered Handle(RegisterPlayer cmd, PlayerState state)
    {
        // Guard
        if (state.Exists)
            throw CommandRejectedError.PreconditionFailed("Player already exists");

        // Validate
        if (string.IsNullOrEmpty(cmd.DisplayName))
            throw CommandRejectedError.InvalidArgument("display_name is required");
        if (string.IsNullOrEmpty(cmd.Email))
            throw CommandRejectedError.InvalidArgument("email is required");

        // Compute
        return new PlayerRegistered
        {
            DisplayName = cmd.DisplayName,
            Email = cmd.Email,
            PlayerType = cmd.PlayerType,
            AiModelId = cmd.AiModelId,
            RegisteredAt = Timestamp.FromDateTime(DateTime.UtcNow),
        };
    }
}

File: examples/cpp/player/agg/handlers/register_handler.cpp
#include "register_handler.hpp"
#include "angzarr/errors.hpp"
#include <chrono>

namespace player {
namespace handlers {

examples::PlayerRegistered handle_register(const examples::RegisterPlayer& cmd, const PlayerState& state) {
    // Guard: preconditions on state
    if (state.exists()) {
        throw angzarr::CommandRejectedError::precondition_failed("Player already exists");
    }

    // Validate: command inputs
    if (cmd.display_name().empty()) {
        throw angzarr::CommandRejectedError::invalid_argument("display_name is required");
    }
    if (cmd.email().empty()) {
        throw angzarr::CommandRejectedError::invalid_argument("email is required");
    }

    // Compute: build event
    examples::PlayerRegistered event;
    event.set_display_name(cmd.display_name());
    event.set_email(cmd.email());
    event.set_player_type(cmd.player_type());
    event.set_ai_model_id(cmd.ai_model_id());

    // Set timestamp
    auto now = std::chrono::system_clock::now();
    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    event.mutable_registered_at()->set_seconds(seconds);

    return event;
}

} // namespace handlers
} // namespace player
